src/Utils/CustomHook.js

```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState, useMemo } from "react";
+import { useSelector, useDispatch } from "react-redux";
+import { useLocation } from "react-router";
+
+import { getAlbums } from "../App/features/albums/albumSlice";
+import {
+  updateSavedUserDetails,
+  updateSavedAlbums,
+  updateSavedSongs,
+} from "../App/features/User/userSlice";
+import { setAllSongs } from "../App/features/allSongs/allSongsSlice";
+import { getFromLocalStorage, saveToLocalStorage } from "./utils";
+
+import axios from "axios";
+
+export default function useAlbums() {
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(getAlbums());
+  }, []);
+}
+
+export function useUserData() {
+  const dispatch = useDispatch();
+  const savedUserDetails = getFromLocalStorage("authUserDetails", {});
+  const savedUserAlbums = getFromLocalStorage("authUserAlbums", []);
+  const savedUserSongs = getFromLocalStorage("authUserSongs", []);
+  const { name, email, token, savedAlbums, savedSongs } = useSelector(
+    (state) => state.user
+  );
+  const isLoggedIn = token ? true : false;
+  const userDetails = useMemo(
+    () => ({
+      name,
+      email,
+      token,
+      isLoggedIn,
+    }),
+    [name, email, token, isLoggedIn]
+  );
+
+  const memoizedSavedAlbums = useMemo(() => savedUserAlbums, [savedUserAlbums]);
+  const memoizedSavedSongs = useMemo(() => savedUserSongs, [savedUserSongs]);
+
+  useEffect(() => {
+    if (Object.keys(savedUserDetails).length > 0 && savedUserDetails.token) {
```

Comment:
use ? before length, example : Object.keys(savedUserDetails)?.length  

change this on other places as well.